File: GraphAlgorithms/Week1/connected_components/ConnectedComponents.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;
import java.util.Scanner;

public class ConnectedComponents {
    private static int numberOfComponents(ArrayList<Integer>[] adj) {
        //write your code here
        return dfs(adj);
    }

    private static int dfs(ArrayList<Integer>[] adj) {
      int cc = 0;
      Set<Integer> set = new HashSet<Integer>();
      for(int i = 0; i < adj.length; i++){
        if(!set.contains(i)) {
          explore(adj, i, set);
          cc++;
        }
      }
      return cc;
    }

    private static void explore(ArrayList<Integer>[] adj, int x, Set<Integer> set) {
      set.add(x);
      for(int i = 0; i < adj[x].size();i++) {
        if(!set.contains(adj[x].get(i))) {
          explore(adj, adj[x].get(i), set);
        }
      }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
            adj[y - 1].add(x - 1);
        }
        System.out.println(numberOfComponents(adj));
    }
}

File: GraphAlgorithms/Week1/connected_components/connected_components.py
#Uses python3

import sys


def number_of_components(adj):
    result = 0
    #write your code here
    visited = set()
    for i in range(len(adj)):
        if i not in visited:
            explore(adj, i, visited)
            result += 1
    return result

def explore(adj, x, visited):
    visited.add(x);
    for i in range(len(adj[x])):
        if adj[x][i] not in visited:
            explore(adj, adj[x][i], visited)

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)
    print(number_of_components(adj))
